Mathlib/Geometry/Manifold/Orientable.lean

```diff
@@ -0,0 +1,192 @@
+- `OrientationReversing.differentiableAt` : an orientation reversing map is differentiable.
+- `orientationPreserving_comp` : a composition between two orientation preserving maps is
+  orientation preserving.
+- `orientationReversing_comp_orientationPreserving` : a composition between an orientation
+  reversing map and an orientation preserving map is orientation reversing.
+- `orientationPreserving_comp_orientationReversing` : a composition between an orientation
+  preserving map and an orientation reversing map is orientation reversing.
+- `orientationReversing_comp` : a composition between two orientation reversing maps is
+  orientation preserving.
+
+## TODO
+
+- Generalize this discussion to any non-trivially normed field.
+- On a given connected set, a diffeomorphism is either orientation preserving or orientation
+  reversing.
+- A normed space (with the trivial model) is orientable.
+- The `n`-sphere is orientable.
+- Products of orientable manifolds are orientable.
+- Define orientations of a smooth manifold, and show that a manifold is orientable if and only if it
+  admits an orientation.
+- Define orientation preserving and reserving maps between manifolds.
+
+## Implementation notes
+
+The current definitions work for differentiable manifolds. For topological manifolds, orientability
+can be defined using *local* orientations (which mathlib cannot state yet, as there is no e.g.
+singular homology). In the future, it would be nice to generalise these definitions to allow for
+topological manifolds also, and relate them to the current definition.
+
+-/
+
+variable {H : Type*} [NormedAddCommGroup H] [NormedSpace ℝ H]
+
+section OrientationPreserving
+
+/--
+A map between normed spaces is orientation-preserving on a given set if it is differentiable and the
+determinant of its Jacobian is strictly positive on that set.
+-/
+def OrientationPreserving (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, 0 < (fderiv ℝ f x).det
+
+/--
+A map between normed spaces is orientation-reversing on a given set if it is differentiable and the
+determinant of its Jacobian is strictly negative on that set.
+-/
+def OrientationReversing (f : H → H) (s : Set H) : Prop :=
+  ∀ x ∈ s, (fderiv ℝ f x).det < 0
+
+lemma orientationPreserving_of_zero_dim (f : H → H) (s : Set H) (h : Module.rank ℝ H = 0) :
+    OrientationPreserving f s :=
+  fun _ _ ↦
+  have det_eq_one : (fderiv ℝ f _).det = 1 := by
+    have b : Basis (Fin 0) ℝ H := Basis.ofEquivFun (finDimVectorspaceEquiv 0 h)
+    rw [ContinuousLinearMap.det, ← (fderiv ℝ f _).det_toMatrix b]
+    exact Matrix.det_fin_zero
+  det_eq_one ▸ Real.zero_lt_one
+
+lemma OrientationPreserving.differentiableAt [FiniteDimensional ℝ H] {f : H → H} {s : Set H}
+    (h : OrientationPreserving f s) {x : H} (hs : x ∈ s) : DifferentiableAt ℝ f x := by
+  cases subsingleton_or_nontrivial H
+  · apply DifferentiableWithinAt.differentiableAt
```

Comment:
The first case can be golfed a little, to
```
apply ((s.subsingleton_of_subsingleton).differentiableOn _ hs).differentiableAt
exact mem_nhds_discrete.mpr hs
```
I would argue this is mathematically "trivial", hence fine.